### PR TITLE
Document --coverage-ignore-choice

### PR DESCRIPTION
**Pull request by dylant-da**

### CODE REVIEW

**Comment by garyverhaegen-da** [comment]
shouldn't this also target 2.8?

**Comment by marciw-da** [comment]
Is this ready for review by daml-docs now? (I had started a review but then you applied more wording improvements -- which is great! -- so just making sure you're done with that sort of thing.)

**Comment by dylant-da** [comment]
I'm done now - sorry, I was worried this would happen :P

**Comment by marciw-da** [comment]
> I'm done now - sorry, I was worried this would happen :P

No problem! Feel free to make more changes -- I won't get to this until a bit later today anyway

File: docs/2.7.2/docs/daml/intro/12_Testing.rst [hunk 1]
```diff
@@ -194,6 +194,8 @@ Flags Controlling Report
 
 Enabling ``--show-coverage`` tells the final printed report to include the names of any templates, choices, and interfaces which are not covered. By default, the report only reports the percentage of coverage.
 
+You can remove choices from the rendered coverage report with ``--coverage-ignore-choice PATTERN``. This flag's behaviour is further documented in `Excluding Choices from the Coverage Report`_.
```
Comment (loiskilberg-da) [inline docs/2.7.2/docs/daml/intro/12_Testing.rst:197]:
```suggestion
You can remove choices from the rendered coverage report with ``--coverage-ignore-choice PATTERN``. This flag's behavior is further documented in `Excluding Choices from the Coverage Report`_.
```

File: docs/2.7.2/docs/daml/intro/12_Testing.rst [hunk 2]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
       --load-coverage test2-results \
       --load-coverage test3-results
 
+.. _Excluding Choices from the Coverage Report:
+
+Excluding Choices from the Coverage Report
+------------------------------------------
+
+Sometimes we would like to exclude choices from the printed coverage report because we aren't interested in them. To do this, we can use the ``--coverage-ignore-choice PATTERN``. Any choice whose fully qualified name matches the regular expression in ``PATTERN`` will be removed from the coverage report. This means that the choice will not be included in counts of defined choices, nor in counts of exercised choices - it will be as if that choice did not exist.
```
Comment (loiskilberg-da) [inline docs/2.7.2/docs/daml/intro/12_Testing.rst:596]:
```suggestion
To exclude choices from the printed coverage report, use ``--coverage-ignore-choice PATTERN``. Any choice whose fully qualified name matches the regular expression in ``PATTERN`` is removed from the coverage report. The choice will not be included in counts of defined choices or in counts of exercised choices. The choice is treated as if it does not exist.
```

File: docs/2.7.2/docs/daml/intro/12_Testing.rst [hunk 3]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+The fully qualified name of a choice is different depending on whether the choice is defined in the local package or in an external package. Choices defined in the local package are fully qualified as ``<module>:<template>:<choice name>``. Choices defined in external packages are fully qualified as ``<package id>:<module>:<template>:<choice name>``. By defining our pattern to match different sections in the fully qualified names of our choices, we can exclude choices based on package id, module, template, or name.
```
Comment (loiskilberg-da) [inline docs/2.7.2/docs/daml/intro/12_Testing.rst:598]:
```suggestion
The fully qualified name of a choice depends on whether the choice is defined in the local package or in an external package. Choices defined in the local package are fully qualified as ``<module>:<template>:<choice name>``. Choices defined in external packages are fully qualified as ``<package id>:<module>:<template>:<choice name>``. By defining your pattern to match different sections in the fully qualified names of your choices, you can exclude choices based on package id, module, template, or name.
```

File: docs/2.7.2/docs/daml/intro/12_Testing.rst [hunk 4]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+Example: Excluding Archive Choices
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To exclude the ``Archive`` choice from coverage, we will want to match for the string "Archive" in the "name" portion of our fully qualified name. We can do this by specifying ``--coverage-ignore-choice ':Archive$'``.
```
Comment (loiskilberg-da) [inline docs/2.7.2/docs/daml/intro/12_Testing.rst:603]:
```suggestion
To exclude the ``Archive`` choice from coverage, match for the string "Archive" in the "name" portion of the fully qualified name. Do this by specifying ``--coverage-ignore-choice ':Archive$'``.
```

File: docs/2.7.2/docs/daml/intro/12_Testing.rst [hunk 5]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+If applied to the coverage report in `Template choice exercise coverage`_, our coverage report changes from the following:
```
Comment (loiskilberg-da) [inline docs/2.7.2/docs/daml/intro/12_Testing.rst:605]:
```suggestion
If applied to the coverage report in `Template choice exercise coverage`_, your coverage report changes from the following:
```

File: docs/2.7.2/docs/daml/intro/12_Testing.rst [hunk 6]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+  internal template choices never exercised: 3
+    Token_Coverage_Part1:T2:C_T2_2
+    Token_Coverage_Part1:T3:C_T3_1
+    Token_Coverage_Part1:T3:C_T3_2
+  ...
+
+Example: Excluding Choices from a Specific Module
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To exclude a specific module (for example ``MyModule``) from coverage, we will want to match for the "module" portion of our fully qualified name. We can do this by specifying ``--coverage-ignore-choice '(^|:)MyModule:[^:]*:[^:]*$'`` - this matches for any template and any choice, matches for our module name, and ignores any leading package identifier.
```
Comment (loiskilberg-da) [inline docs/2.7.2/docs/daml/intro/12_Testing.rst:640]:
```suggestion
To exclude a specific module (for example ``MyModule``) from coverage, match for the "module" portion of the fully qualified name. Do this by specifying ``--coverage-ignore-choice '(^|:)MyModule:[^:]*:[^:]*$'``. This matches for any template and any choice, matches for your module name, and ignores any leading package identifier.
```

File: docs/2.7.2/docs/daml/intro/12_Testing.rst [hunk 7]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+Excluding Choices from Serialized Reports
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To ensure that serialized data always reflects full coverage information, the flag does **not** eliminate the choices from serialization via the ``--save-coverage`` flag. Serialized reports saved to a file always contain all information collected -- the ``--coverage-ignore-choice`` flag only excludes choices from the printed report. This means any text report generated from serialized data will need to specify ``--coverage-ignore-choice`` every time it is generated.
```
Comment (loiskilberg-da) [inline docs/2.7.2/docs/daml/intro/12_Testing.rst:645]:
```suggestion
To ensure that serialized data always reflects full coverage information, the flag does **not** eliminate the choices from serialization using the ``--save-coverage`` flag. Serialized reports saved to a file always contain all information collected. The ``--coverage-ignore-choice`` flag only excludes choices from the printed report. For any text report generated from serialized data, you must specify ``--coverage-ignore-choice`` every time it is generated.
```

File: docs/2.8.0/docs/daml/intro/12_Testing.rst [hunk 8]
```diff
@@ -194,6 +194,8 @@ Flags Controlling Report
 
 Enabling ``--show-coverage`` tells the final printed report to include the names of any templates, choices, and interfaces which are not covered. By default, the report only reports the percentage of coverage.
 
+You can remove choices from the rendered coverage report with ``--coverage-ignore-choice PATTERN``. This flag's behaviour is further documented in `Excluding Choices from the Coverage Report`_.
```
Comment (loiskilberg-da) [inline docs/2.8.0/docs/daml/intro/12_Testing.rst:197]:
```suggestion
You can remove choices from the rendered coverage report with ``--coverage-ignore-choice PATTERN``. This flag's behavior is further documented in `Excluding Choices from the Coverage Report`_.
```

File: docs/2.8.0/docs/daml/intro/12_Testing.rst [hunk 9]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
       --load-coverage test2-results \
       --load-coverage test3-results
 
+.. _Excluding Choices from the Coverage Report:
+
+Excluding Choices from the Coverage Report
+------------------------------------------
+
+Sometimes we would like to exclude choices from the printed coverage report because we aren't interested in them. To do this, we can use the ``--coverage-ignore-choice PATTERN``. Any choice whose fully qualified name matches the regular expression in ``PATTERN`` will be removed from the coverage report. This means that the choice will not be included in counts of defined choices, nor in counts of exercised choices - it will be as if that choice did not exist.
```
Comment (loiskilberg-da) [inline docs/2.8.0/docs/daml/intro/12_Testing.rst:596]:
```suggestion
To exclude choices from the printed coverage report, use ``--coverage-ignore-choice PATTERN``. Any choice whose fully qualified name matches the regular expression in ``PATTERN`` is removed from the coverage report. The choice will not be included in counts of defined choices or in counts of exercised choices. The choice is treated as if it does not exist.
```

File: docs/2.8.0/docs/daml/intro/12_Testing.rst [hunk 10]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+The fully qualified name of a choice is different depending on whether the choice is defined in the local package or in an external package. Choices defined in the local package are fully qualified as ``<module>:<template>:<choice name>``. Choices defined in external packages are fully qualified as ``<package id>:<module>:<template>:<choice name>``. By defining our pattern to match different sections in the fully qualified names of our choices, we can exclude choices based on package id, module, template, or name.
```
Comment (loiskilberg-da) [inline docs/2.8.0/docs/daml/intro/12_Testing.rst:598]:
```suggestion
The fully qualified name of a choice depends on whether the choice is defined in the local package or in an external package. Choices defined in the local package are fully qualified as ``<module>:<template>:<choice name>``. Choices defined in external packages are fully qualified as ``<package id>:<module>:<template>:<choice name>``. By defining your pattern to match different sections in the fully qualified names of your choices, you can exclude choices based on package id, module, template, or name.
```

File: docs/2.8.0/docs/daml/intro/12_Testing.rst [hunk 11]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+Example: Excluding Archive Choices
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To exclude the ``Archive`` choice from coverage, we will want to match for the string "Archive" in the "name" portion of our fully qualified name. We can do this by specifying ``--coverage-ignore-choice ':Archive$'``.
```
Comment (loiskilberg-da) [inline docs/2.8.0/docs/daml/intro/12_Testing.rst:603]:
```suggestion
To exclude the ``Archive`` choice from coverage, match for the string "Archive" in the "name" portion of the fully qualified name. Do this by specifying ``--coverage-ignore-choice ':Archive$'``.
```

File: docs/2.8.0/docs/daml/intro/12_Testing.rst [hunk 12]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+If applied to the coverage report in `Template choice exercise coverage`_, our coverage report changes from the following:
```
Comment (loiskilberg-da) [inline docs/2.8.0/docs/daml/intro/12_Testing.rst:605]:
```suggestion
If applied to the coverage report in `Template choice exercise coverage`_, your coverage report changes from the following:
```

File: docs/2.8.0/docs/daml/intro/12_Testing.rst [hunk 13]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+  internal template choices never exercised: 3
+    Token_Coverage_Part1:T2:C_T2_2
+    Token_Coverage_Part1:T3:C_T3_1
+    Token_Coverage_Part1:T3:C_T3_2
+  ...
+
+Example: Excluding Choices from a Specific Module
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To exclude a specific module (for example ``MyModule``) from coverage, we will want to match for the "module" portion of our fully qualified name. We can do this by specifying ``--coverage-ignore-choice '(^|:)MyModule:[^:]*:[^:]*$'`` - this matches for any template and any choice, matches for our module name, and ignores any leading package identifier.
```
Comment (loiskilberg-da) [inline docs/2.8.0/docs/daml/intro/12_Testing.rst:640]:
```suggestion
To exclude a specific module (for example ``MyModule``) from coverage, match for the "module" portion of the fully qualified name. Do this by specifying ``--coverage-ignore-choice '(^|:)MyModule:[^:]*:[^:]*$'``. This matches for any template and any choice, matches for your module name, and ignores any leading package identifier.
```

File: docs/2.8.0/docs/daml/intro/12_Testing.rst [hunk 14]
```diff
@@ -584,6 +588,62 @@ If a test failure causes one ``daml test`` to fail, other coverage results from
...
+
+Excluding Choices from Serialized Reports
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To ensure that serialized data always reflects full coverage information, the flag does **not** eliminate the choices from serialization via the ``--save-coverage`` flag. Serialized reports saved to a file always contain all information collected -- the ``--coverage-ignore-choice`` flag only excludes choices from the printed report. This means any text report generated from serialized data will need to specify ``--coverage-ignore-choice`` every time it is generated.
```
Comment (loiskilberg-da) [inline docs/2.8.0/docs/daml/intro/12_Testing.rst:645]:
```suggestion
To ensure that serialized data always reflects full coverage information, the flag does **not** eliminate the choices from serialization using the ``--save-coverage`` flag. Serialized reports saved to a file always contain all information collected. The ``--coverage-ignore-choice`` flag only excludes choices from the printed report. For any text report generated from serialized data, you must specify ``--coverage-ignore-choice`` every time it is generated.
```

**Review by loiskilberg-da — COMMENTED** [review]
Made some language suggestions (taking over from Marci).

**Review by loiskilberg-da — APPROVED** [review]
Looks good!

**Comment by garyverhaegen-da** [comment]
Rebased on https://github.com/digital-asset/docs.daml.com/pull/469.

**Pull request merged**